### DISCUSSION TITLE
Issues with TLS/SSL (If your app stopped working Read Me First!!!)

### DISCUSSION THREAD

**Discussion by brandlink-james** [mozts2005/ZendeskApi_v2#339] [Q&A] [answered]
Today the api does not work and the error 'The underlying connection was closed: An unexpected error occurred on a send' pops up when executing a simple query
api.Users.Search();

I also use the postman to do the same query and it still works well. 

Is there anything wrong with this library?

**Comment by robinschrijer** [comment]
We are having the same issue. Same error,. Our last successful ticket via the API was 19 jun. 2018 22:35:55.

**Comment by yonahs** [comment]
I am having similar issues except that I have discovered that my issues are not just with zendesk but with a bunch of my API calls

**Comment by mozts2005** [comment, marked as answer]
Zendesk is only[ supporting TLS 1.2](https://support.zendesk.com/hc/en-us/articles/360000710347-Removal-of-support-for-TLS-protocol-versions-1-0-and-1-1) as of June 13, 2018

Options to fix this issue (in the order I would do them) 

1. update your .NET project to .NET 4.7.X
1.  add following code to your application this code will affect all HTTP calls.
This setting is global and may need to be different for other services. 
```Csharp
System.Net.ServicePointManager.SecurityProtocol = System.Net.SecurityProtocolType.Tls12 | System.Net.SecurityProtocolType.Tls11 | System.Net.SecurityProtocolType.Tls;
```
note:
https://docs.microsoft.com/en-us/dotnet/framework/network-programming/tls#support-for-tls-12

**Comment by brandlink-james** [comment]
Hey Mozts, the 2nd solution works like a charm. 

I cannot do the 1st solution as I cannot install 4.7.x in my computer which may be caused by the existing .Net core version in my system. 

Anyway the 2nd does save my day. Thanks:)

**Comment by robinschrijer** [comment]
Hi Mozts, i also used the second option. Works great.

**Comment by anmiles** [comment]
Hi Mozts, 
I have installed 4.7.2 to staging server and changed target framework of my project to 4.7.2 and built it again.
Nothing changed.
So I added a line of code from 2nd option and error seems to be gone!
Thank you!

**Comment by brandlink-james** [comment]
Hi Mozts,

I am not sure if the ServicePointManager.SecurityProtocol will affect other service/connection. Could we add the setting only at this library?

How about change a little bit to 

System.Net.ServicePointManager.SecurityProtocol |= System.Net.SecurityProtocolType.Tls12 | System.Net.SecurityProtocolType.Tls11 | System.Net.SecurityProtocolType.Tls;

Use |= instead of =

**Comment by mozts2005** [comment]
@brandlink-james As I pointed out before  Microsoft says not do so please see https://docs.microsoft.com/en-us/dotnet/framework/network-programming/tls 

System.Net.ServicePointManager.SecurityProtocol is a static as can be seen here 
https://github.com/Microsoft/referencesource/blob/master/System/net/System/Net/ServicePointManager.cs#L249

If I change it in the library, this could brake calls you make to other services (say one that errors on a tls1.2 call). I have had to fix that issue for my day job.

**Comment by rwjdk** [comment]
Just for reference, this is breaking all the unit tests for the project at the moment.

Easiest way to fix this is to add this class outside any namespace (either in new file or at the end of Settings.cs perhaps - can't be in a namespace)

```
[SetUpFixture]
public class InitializeTests
{
    [OneTimeSetUp]
    public void OneTimeSetUp()
    {
        ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12 | SecurityProtocolType.Tls11 | SecurityProtocolType.Tls;
    }
}
```

@mozts2005: Want a pull request for this or add yourself (sorry, if I ask to many question. Newbe when it come to open source contributions)
